Add function to remove duplicate rows in a Table

evertrol opened the pull request:
This function is similar to table.operations.unique, except that it
can remove *all* duplicate rows, thus only keeping rows that never had
a duplicate in the input table. [#4617]

taldcroft commented:
@evertrol - Nice.  

What about just adding the `keep` kwarg to `unique` (and in effect replacing the existing `unique` with your more generalized method).  The existing docstring for `unique` is actually slightly ambiguous in just the right way -- "Returns the unique rows of a table" could be interpreted as returning only the rows of a table that are unique.

I would also suggest having the allowed values of `keep` be `first`, `last` and `none` for better symmetry.

taldcroft commented on astropy/table/operations.py:
```diff
@@ -380,6 +380,68 @@ def unique(input_table, keys=None, silent=False):
+        cleaned_table = grouped_table[grouped_table.groups.indices[:-1]]
+    elif keep == 'last':
+        cleaned_table = grouped_table[grouped_table.groups.indices[1:]-1]
+    else:
```
Explicitly check `elif keep == 'none':` and then have a final `else` that raises a ValueError exception for an invalid input.

evertrol replied:
I prefer to do the keyword validity check near the start of the function, together with the handling of the other parameters. Not sure if there is a clear preference here, technically or astropy wise.
I've added 'none' (and `None`) as valid values, though `False` and `None` are not documented and just silently accepted (may be convenient, could be a bad thing).

taldcroft commented on astropy/table/operations.py:
```diff
@@ -380,6 +380,68 @@ def unique(input_table, keys=None, silent=False):
+    elif keep == 'last':
+        cleaned_table = grouped_table[grouped_table.groups.indices[1:]-1]
+    else:
+        cleaned_table = vstack([t for t in grouped_table.groups
```
This works but will be wicked slow.  You should be able to do this in a vectorized way taking advantage of `np.diff` and boolean masks to generate an array of indices to then use with `grouped_table[...]`.

taldcroft replied:
Structurally you can do this with less repetition and no intermediate `cleaned_table` with:
```
if keep == 'first':
    indices = grouped_table.groups.indices[-1]
elif keep == 'last':
    indices = ...
elif keep == 'none':
    ...
else:
    raise ValueError(..)

return grouped_table[indices]
```

evertrol replied:
Done. I tend to get confused when doing tricks with `np.diff` and boolean masks (since `np.diff` loses one element), so I hope I covered all edge cases in the tests. And indeed, a pretty major speed increase this way.

evertrol commented:
@taldcroft Thanks for having a look.
I made it a separate function instead of extending `unique()`, since the name of `unique` doesn't really indicate the extra behaviour one gets with `keep=False|'none'|None`, even with an updated doc string.
Perhaps a new function is too much for some small added functionality, but I like the clarity of it.

For now, I've updated the PR so that `unique` calls `remove_duplicate_rows`.

taldcroft commented:
I think that the original `unique` benefits from the new `keep` arg by way of allowing a choice for which row to keep in the case of duplicates (first or last).  So at that point the only thing `remove_duplicate_rows` is providing over `unique` is literally removing all rows that are duplicates instead of de-duplicating.  That seems too small a use-case to have its own method.  

I think that anyone who is bothered by duplicate rows in their table will at least have a look at `unique` to see what it can do, and they will be happy to see the `keep` arg is exactly what they need!

taldcroft commented:
About the allowed values for `keep`, I would prefer to not have values that are not documented but silently do something.  This allows for buggy code to run and also makes it harder to read code.  For instance if I see some code `du = dat.unique(keep=False)` and go to the documentation and do not see that listed as an option, then I have to *guess* what is the intended result.

Please just have the three string options and don't accept anything else.

taldcroft commented:
@evertrol - you could also add a documentation example with relevant words that will get picked up by search engines, and direct people to investigate `unique`.

evertrol commented:
I agree with having `keep` no (hidden) extra values.
I've now removed `remove_duplicate_rows()` and added the keyword to `unique()` instead. 
And included a bunch of examples in the doc string.

evertrol commented:
Perhaps confusingly so, I've turned this (accidentally) into a new PR, since I renamed the branch and commit (as the original title doesn't seem fitting anymore). See PR #4632
